src/hotspot/share/utilities/zipLibrary.cpp

```diff
@@ -59,39 +59,42 @@ static inline bool not_loaded() {
   return !is_loaded();
 }
 
-static void* dll_lookup(const char* name, const char* path) {
+static void* dll_lookup(const char* name, const char* path, bool vm_exit_on_failure) {
   assert(_zip_handle != nullptr, "invariant");
   void* func = os::dll_lookup(_zip_handle, name);
-  if (func == nullptr) {
+  if (func == nullptr && vm_exit_on_failure) {
```

Comment:
This seems to be unraveling somewhat. If we don't exit-on-failure then the callers have to be prepared for that failure and it isn't obvious to me that is always the case. And this is no longer just a refactoring. So I think a separate issue would need to be filed to look at this change in behaviour.

Reply:
I understand your point. The complication comes about from the usage in HeadDumperCompression, which only wants to reuse some functions in the zip library, init_params() and compress(). This caller is only invoked by means of an externally invoked jcmd, HeapDumpDCmd, iff the gzip compression level is set. To have this jcmd exit the VM is a departure from existing behaviour. So this change aligns with the existing behavior.

The other client, which uses this in a more structured way, is ClassLoader:. It is used also by the JNI entry point JVM_LoadZipLibrary(). Even from this JNI entry point, the behaviour is to perform a vm_exit on failure.

Reply:
My current take is to keep the existing behaviours of these two callers in order to get rid of the Zip_lock.

Should there be more callers in the future, additional adjustments and refactorings could be performed at that time.

Reply:
I'd rather see this PR just address the refactoring and locking changes without changing any actual behaviour in terms of "vm exit" - that seems a separate issue.

Reply:
That would mean committing a behaviour change. It is it not a separate issue.

Reply:
Sorry I don't follow. Anything that currently does vm_exit should continue to do so; and anything that does not should continue to do not. The current code does not appear to follow that.

Reply:
It follows that now, with the introduction of the vm_exit construct.

Reply:
Sorry I am having a hard time seeing the equivalence. The original code doesn't potentially exit after a failed dll_lookup AFAICS - only a failed dl_load does that.

Reply:
The original code in ClassLoader does not check that the resolved method entry points through dll_lookup are found or resolved correctly. As there are no null checks before using the function pointers, it would instantly crash if the entry points were not resolved correctly. This is semantically equivalent.

The only difference is that ClassLoader, should it be the initializer of the library, also attempts to resolve the init_params() and compress() entry points (used by HeapDumpCompressor).

Reply:
Adjusted to never exit the vm if failing to resolve InitParams and Compress.

Reply:
I think the original code worked on the basis that we know what a libzip is and provides and that is not likely to change, so all the expected entry points will be present.

Reply:
Then I don't know why HeapDumperCompression behaves differently.